Integrate e2e for CI

LeonidBeynenson opened the pull request:
Allows running tests/test_ote_api.py on internal CI.
Note that this change should be transparent; that is inside environment without
installed e2e-test-framework all tests should work in the same manner as earlier.

Note that tests/test_ote_training.py is not ready at the moment and should not be used yet.

Other changes:
* in mmdet/apis/export.py -- automatically choosing Model Optimizer command 
  between "mo" amd "mo.py"
* in mmdet/apis/ote/apis/detection/config_utils.py -- improve calculation of interval
  between evaluations for small number of iterations.

druzhkov-paul commented on configs/ote/custom-object-detection/mobilenetV2_ATSS/model.py:
```diff
@@ -75,6 +75,6 @@
 log_level = 'INFO'
 work_dir = 'output'
-load_from = None
+load_from = '/usr/src/app/external/mmdetection/ote_data/MODELS/mobilenetV2_ATSS/coco_snapshot.pth'
```
What is this?

LeonidBeynenson replied:
It is the path of the model snapshot in CI.
We did not implement a mechanism that patches/substitutes the file model.py for tests, so wrote the test paths right inside the model.py

druzhkov-paul replied:
Bad news is that it'll break everything. You cannot use the model with invalid/unreachable path in `load_from` config parameter. Please come up with a different approach here.

LeonidBeynenson replied:
Is it possible to push the weights of the model to a server as for other models? It will resolve all the issues.

Another option: we could use environment variables:
```python
import os
model_path = os.environ.get('OTE_TEST_MODEL_PATH')
if model_path is not None:
    load_from = os.path.join(model_path, 'mobilenetV2_ATSS/coco_snapshot.pth')
else:
    load_from = None
```
Note that this will be a temporal decision until the weights are loded to the server.

The third option is to patch the model.py before loading. In this case we should load `OTE_TEST_MODEL_PATH` in test_ote_api.py, create temporary folder, copy all the files from the `template_dir` to the temporary folder, and patch model.py in it. 
But it will break this relative path to `configuration.yaml` in `template.yaml`:
```yaml
hyper_parameters:
  base_path: ../../../../mmdet/apis/ote/apis/detection/configuration.yaml
```
To avoid breaking the relative paths I should 
* either copy the whole git folder to the temporary folder just to patch `model.py` 
* or create the temporary folder just inside the git folder during the test

All these options are a bit weird (except the first one, but it is the longest), so you could propose what should be done here.

LeonidBeynenson replied:
The option 2 is implemented and it works.

druzhkov-paul commented on configs/ote/face-detection/face-detection-0200/model.py:
```diff
@@ -1,15 +1,16 @@
+_base_ = [
+    './coco_data_pipeline.py'
+]
```
No need to keep data pipeline in a separate file anymore.

I also wonder, why have you changed the data pipeline at all?

LeonidBeynenson replied:
We made it w.r.t. the changes that were in model.py earlier

LeonidBeynenson replied:
This file is removed from the current PR

druzhkov-paul commented on configs/ote/face-detection/face-detection-0200/template.yaml:
```diff
@@ -0,0 +1,101 @@
+# Would be better to have something more precise and self-contained, e.g.:  docker image or path to Python venv activation script?
+framework: OTEDetection v2.9.1
+
+dependencies:
+- source: ../../../../mmdet/apis/ote/apis/detection/configuration.yaml
+  destination: configuration.yaml
```
`dependencies` is deprecated.

LeonidBeynenson replied:
I know, did not merge changes for the template files yet

LeonidBeynenson replied:
These files are removed from the current PR.

druzhkov-paul commented on configs/ote/face-detection/face-detection-0200/template.yaml:
```diff
@@ -0,0 +1,101 @@
+
+# Task and hyperparameters.
+hyper_parameters:
+  base_path: configuration.yaml
```
Invalid relative path.

LeonidBeynenson replied:
I know, did not merge changes for the template files yet

LeonidBeynenson replied:
These files are removed from the current PR.

druzhkov-paul commented on configs/ote/template_paths.yml:
```diff
@@ -0,0 +1,26 @@
+'mobilenet_v2-2s_ssd-256x256': ./custom-object-detection/mobilenet_v2-2s_ssd-256x256/template.yaml
+'mobilenet_v2-2s_ssd-384x384': ./custom-object-detection/mobilenet_v2-2s_ssd-384x384/template.yaml
+'mobilenet_v2-2s_ssd-512x512': ./custom-object-detection/mobilenet_v2-2s_ssd-512x512/template.yaml
+'mobilenetV2_ATSS': ./custom-object-detection/mobilenetV2_ATSS/template.yaml
+'mobilenetV2_SSD': ./custom-object-detection/mobilenetV2_SSD/template.yaml
+'resnet50_VFNet': ./custom-object-detection/resnet50_VFNet/template.yaml
+'face-detection-0200': ./face-detection/face-detection-0200/template.yaml
+'face-detection-0202': ./face-detection/face-detection-0202/template.yaml
+'face-detection-0204': ./face-detection/face-detection-0204/template.yaml
+'face-detection-0205': ./face-detection/face-detection-0205/template.yaml
+'face-detection-0206': ./face-detection/face-detection-0206/template.yaml
+'face-detection-0207': ./face-detection/face-detection-0207/template.yaml
+'horizontal-text-detection-0001': ./horizontal-text-detection/horizontal-text-detection-0001/template.yaml
+'person-detection-0200': ./person-detection/person-detection-0200/template.yaml
+'person-detection-0201': ./person-detection/person-detection-0201/template.yaml
+'person-detection-0202': ./person-detection/person-detection-0202/template.yaml
+'person-detection-0203': ./person-detection/person-detection-0203/template.yaml
+'person-vehicle-bike-detection-2000': ./person-vehicle-bike-detection/person-vehicle-bike-detection-2000/template.yaml
+'person-vehicle-bike-detection-2001': ./person-vehicle-bike-detection/person-vehicle-bike-detection-2001/template.yaml
+'person-vehicle-bike-detection-2002': ./person-vehicle-bike-detection/person-vehicle-bike-detection-2002/template.yaml
+'person-vehicle-bike-detection-2003': ./person-vehicle-bike-detection/person-vehicle-bike-detection-2003/template.yaml
+'person-vehicle-bike-detection-2004': ./person-vehicle-bike-detection/person-vehicle-bike-detection-2004/template.yaml
+'vehicle-detection-0200': ./vehicle-detection/vehicle-detection-0200/template.yaml
+'vehicle-detection-0201': ./vehicle-detection/vehicle-detection-0201/template.yaml
+'vehicle-detection-0202': ./vehicle-detection/vehicle-detection-0202/template.yaml
+'vehicle-detection-0203': ./vehicle-detection/vehicle-detection-0203/template.yaml
```
What's the reason to have that kind of list?

LeonidBeynenson replied:
This file is required for `test_ote_training.py` file. It is pointed as a pytest parameter `--template-paths` (see the file `tests/conftest.py`). The parameter allows to point for a model name the realtive path where template file is placed.

Note that during tests development we used other `template_paths.yml` files as the parameter.

LeonidBeynenson replied:
This file is removed from the current PR.

druzhkov-paul replied:
Such kind of lists are error-prone and hard to support. At the same time it does not match the way it's done in OTE, which might cause another point of confusion.

LeonidBeynenson replied:
Ok! At the moment I remove this file from this PR.
I think I will fix this your comment when I make next PR with my `test_ote_training.py`.

druzhkov-paul commented on constraints.txt:
```diff
@@ -19,3 +19,8 @@ flatbuffers~=1.12.0
+torch>=1.7.1,<=1.9.0
+torchvision>=0.8.1,<=0.10.0
+
+# mmcv-full==1.3.12
```
Is it safe to remove this?

LeonidBeynenson replied:
Yes, will do that

LeonidBeynenson replied:
Fixed

druzhkov-paul commented on constraints.txt:
```diff
@@ -19,3 +19,8 @@ flatbuffers~=1.12.0
 grpcio-health-checking==1.32.0
 six~=1.15.0
+
+torch>=1.7.1,<=1.9.0
+torchvision>=0.8.1,<=0.10.0
```
So... you also put torch and torchvision versions to constraints file inside init_venv.sh. Why do you need both?

LeonidBeynenson replied:
Ok, will remove these lines in `constraints.txt`

LeonidBeynenson replied:
Done

druzhkov-paul commented on mmdet/apis/export.py:
```diff
@@ -124,6 +124,15 @@ def add_node_names(export_name):
 
 
+def _get_mo_cmd():
+    for mo_cmd in ('mo', 'mo.py'):
```
Under what circumstances we end up lacking `mo` command, if we always install it via pip?

LeonidBeynenson replied:
Initially I used the code with OpenVINO 2021.4 installed from a package, so this code was required.
Later I switched to `init_venv.sh` and used OpenVINO pip package.

But note that sometimes we still **need** using OpenVINO package. As an example,  if we need to check OpenVINO latest features -- this week I had to get recent OpenVINO package from OpenVINO CI as a set of zip files, extract them, and use `setupvars.sh`
(and yes, after that Model Optimizer is `mo.py` instead of `mo`)

So, I propose to keep this change, since it may be useful.

druzhkov-paul replied:
Looks more like a legacy support effort, but Ok, let's leave it as is.

druzhkov-paul commented on mmdet/apis/ote/apis/detection/config_utils.py:
```diff
@@ -78,8 +79,8 @@ def set_hyperparams(config: Config, hyperparams: OTEDetectionConfig):
     num_checkpoints = int(hyperparams.learning_parameters.num_checkpoints)
-    config.evaluation.interval = total_iterations // num_checkpoints
-    config.checkpoint_config.interval = total_iterations // num_checkpoints
+    config.evaluation.interval = math.ceil(total_iterations / num_checkpoints)
+    config.checkpoint_config.interval = math.ceil(total_iterations / num_checkpoints)
```
What are the pros of it?

LeonidBeynenson replied:
If we use in some micro-training `total_iterations=3`, we could have `total_iterations < num_checkpoints`. In this case `config.evaluation.interval` was `0` and error happened.

druzhkov-paul replied:
I see. Thanks.

druzhkov-paul commented on tests/test_ote_api.py:
```diff
@@ -51,27 +52,36 @@
 from mmdet.apis.ote.apis.detection.config_utils import set_values_as_default
 from mmdet.apis.ote.apis.detection.ote_utils import generate_label_schema
 
+from e2e_test_system import e2e_pytest_api
```
Just wondering if that kind of import works? From what directory should the tests be run?

LeonidBeynenson replied:
Sorry, I do not know. But it works!

druzhkov-paul reviewed (changes requested):
Is it possible to exclude irrelevant changes from the PR (anything relative to non custom object detection models)?

I'd also consider squashing commits before merge.

LeonidBeynenson committed:
Allows running tests/test_ote_api.py on internal CI.

Other changes:
* in mmdet/apis/export.py -- automatically choosing Model Optimizer
  command between "mo" amd "mo.py"
* in mmdet/apis/ote/apis/detection/config_utils.py -- improve calculation
  of interval between evaluations for small number of iterations.

LeonidBeynenson commented:
> Is it possible to exclude irrelevant changes from the PR (anything relative to non custom object detection models)?
> I'd also consider squashing commits before merge.

Excluded
Squashed

druzhkov-paul reviewed (changes requested):
Please resolve all questions from the previous review.

LeonidBeynenson committed:
Temporary decision for load_from in custom-object-detection models' configs

LeonidBeynenson committed:
Debug changes in config files

LeonidBeynenson committed:
Fix issue with 'import os' in mmdetection config

Also some debug insertions are removed.

LeonidBeynenson commented:
2 @druzhkov-paul : Pavel, everything is done.

druzhkov-paul commented:
222 @LeonidBeynenson  I guess "everything" does not include merge conflicts resolution.

LeonidBeynenson committed:
Merge branch 'dp/nous_sdk_api' into lb/integrate_e2e
